File: agent_shared_backend/src/agents/travel_planner/state.py
"""Shared AG-UI state model for the travel planner agent."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .prompts import ClientId, ClientPromptParams, client_prompt_params

WorkspaceStage = Literal["idle", "requirements", "draft", "comparison", "review", "final"]
WorkflowStage = Literal["requirements", "draft", "comparison", "review", "final"]
ApprovalStatus = Literal["not_requested", "pending", "approved", "rejected"]
RequirementStatus = Literal["missing", "draft", "confirmed"]
PolicyStatus = Literal["approved", "warning", "rejected"]
ItineraryCategory = Literal["travel", "meeting", "work", "meal", "lodging", "buffer"]


class RequirementCard(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    value: str
    status: RequirementStatus = "draft"
    description: str | None = None


class ItineraryItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    day: str
    date: str | None = None
    weekday: str | None = None
    title: str
    detail: str
    start_time: str | None = Field(default=None, alias="startTime")
    end_time: str | None = Field(default=None, alias="endTime")
    location: str | None = None
    category: ItineraryCategory = "work"
    notes: list[str] = Field(default_factory=list)


class DraftPlan(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    title: str
    summary: str
    timezone: str = "GMT+09"
    calendar_start_hour: int = Field(default=7, alias="calendarStartHour")
    calendar_end_hour: int = Field(default=22, alias="calendarEndHour")
    itinerary: list[ItineraryItem] = Field(default_factory=list)


class HotelOption(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    name: str
    location: str
    nightly_rate_krw: int = Field(alias="nightlyRateKrw")
    total_krw: int = Field(alias="totalKrw")
    policy_status: PolicyStatus = Field(alias="policyStatus")
    highlights: list[str] = Field(default_factory=list)


class PolicyCheck(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    status: PolicyStatus
    detail: str


class FinalPlan(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    title: str
    summary: str
    approval_note: str = Field(alias="approvalNote")


class TravelPlannerState(BaseModel):
    """State synchronized through AG-UI between CopilotKit and Pydantic AI.

    CopilotKit owns the browser-side copy. Every run sends the current state in
    `RunAgentInput.state`; Pydantic AI validates it here and reinjects it into
    the agent through `RunContext.deps.state`.

    This schema mirrors the CopilotKit frontend workspace contract. It is not a
    backend-only persistence model. Frontend workspace tools mutate these fields
    in the browser agent state, and the next AG-UI request sends the same state
    back here for context reinjection.
    """

    model_config = ConfigDict(extra="allow", populate_by_name=True)

    stage: WorkspaceStage | None = None
    workflow_stage: WorkflowStage = Field(default="requirements", alias="workflowStage")
    destination: str | None = None
    start_date: str | None = Field(default=None, alias="startDate")
    nights: int | None = None
    budget_krw: int | None = Field(default=None, alias="budgetKrw")
    trip_purpose: str | None = Field(default=None, alias="tripPurpose")
    preferences: list[str] = Field(default_factory=list)
    selected_hotel_id: str | None = Field(default=None, alias="selectedHotelId")
    approval_status: ApprovalStatus = Field(default="not_requested", alias="approvalStatus")
    last_user_action: str | None = Field(default=None, alias="lastUserAction")
    pending_action: str | None = Field(default=None, alias="pendingAction")
    requirement_cards: list[RequirementCard] = Field(default_factory=list, alias="requirementCards")
    draft_plan: DraftPlan | None = Field(default=None, alias="draftPlan")
    hotel_options: list[HotelOption] = Field(default_factory=list, alias="hotelOptions")
    policy_checks: list[PolicyCheck] = Field(default_factory=list, alias="policyChecks")
    final_plan: FinalPlan | None = Field(default=None, alias="finalPlan")


@dataclass
class TravelPlannerDeps:
    """Agent dependencies passed to every Pydantic AI run.

    This dataclass intentionally has a non-optional `state` field so it satisfies
    Pydantic AI's `StateHandler` protocol. Without this shape, AG-UI state sent
    by CopilotKit is ignored by `AGUIAdapter`.
    """

    client: ClientPromptParams
    state: TravelPlannerState = field(default_factory=TravelPlannerState)


def travel_planner_deps(client_id: ClientId) -> TravelPlannerDeps:
    return TravelPlannerDeps(client=client_prompt_params(client_id))

File: copilotkit_agui/src/business-trip/state.ts
import { z } from "zod";
import {
  businessTripStages,
  isBusinessTripStage,
  type BusinessTripStage,
} from "@/business-trip/routes";

export const approvalStatusValues = [
  "not_requested",
  "pending",
  "approved",
  "rejected",
] as const;

export const policyStatusValues = ["approved", "warning", "rejected"] as const;
export const requirementStatusValues = ["missing", "draft", "confirmed"] as const;
export const itineraryCategoryValues = [
  "travel",
  "meeting",
  "work",
  "meal",
  "lodging",
  "buffer",
] as const;

export const businessTripStageSchema = z.enum(businessTripStages);
export const approvalStatusSchema = z.enum(approvalStatusValues);
export const policyStatusSchema = z.enum(policyStatusValues);
export const requirementStatusSchema = z.enum(requirementStatusValues);
export const itineraryCategorySchema = z.enum(itineraryCategoryValues);

export const requirementCardSchema = z.object({
  id: z.string(),
  title: z.string(),
  value: z.string(),
  status: requirementStatusSchema.default("draft"),
  description: z.string().nullish(),
});

export const itineraryItemSchema = z.object({
  day: z.string(),
  date: z.string().nullish(),
  weekday: z.string().nullish(),
  title: z.string(),
  detail: z.string(),
  startTime: z.string().nullish(),
  endTime: z.string().nullish(),
  location: z.string().nullish(),
  category: itineraryCategorySchema.default("work"),
  notes: z.array(z.string()).default([]),
});

export const draftPlanSchema = z.object({
  title: z.string(),
  summary: z.string(),
  timezone: z.string().default("GMT+09"),
  calendarStartHour: z.number().int().min(0).max(23).default(7),
  calendarEndHour: z.number().int().min(1).max(24).default(22),
  itinerary: z.array(itineraryItemSchema).default([]),
});

export const hotelOptionSchema = z.object({
  id: z.string(),
  name: z.string(),
  location: z.string(),
  nightlyRateKrw: z.number().int().nonnegative(),
  totalKrw: z.number().int().nonnegative(),
  policyStatus: policyStatusSchema,
  highlights: z.array(z.string()).default([]),
});

export const policyCheckSchema = z.object({
  id: z.string(),
  title: z.string(),
  status: policyStatusSchema,
  detail: z.string(),
});

export const finalPlanSchema = z.object({
  title: z.string(),
  summary: z.string(),
  approvalNote: z.string(),
});

export const businessTripWorkspaceStateSchema = z
  .object({
    stage: z.union([businessTripStageSchema, z.literal("idle")]).nullish(),
    workflowStage: businessTripStageSchema.optional(),
    destination: z.string().nullish(),
    startDate: z.string().nullish(),
    nights: z.number().int().positive().nullish(),
    budgetKrw: z.number().int().positive().nullish(),
    tripPurpose: z.string().nullish(),
    preferences: z.array(z.string()).default([]),
    selectedHotelId: z.string().nullish(),
    approvalStatus: approvalStatusSchema.default("not_requested"),
    lastUserAction: z.string().nullish(),
    pendingAction: z.string().nullish(),
    requirementCards: z.array(requirementCardSchema).default([]),
    draftPlan: draftPlanSchema.nullish(),
    hotelOptions: z.array(hotelOptionSchema).default([]),
    policyChecks: z.array(policyCheckSchema).default([]),
    finalPlan: finalPlanSchema.nullish(),
  });

export const businessTripWorkspacePatchSchema = z
  .object({
    stage: businessTripStageSchema.optional(),
    workflowStage: businessTripStageSchema.optional(),
    destination: z.string().nullish(),
    startDate: z.string().nullish(),
    nights: z.number().int().positive().nullish(),
    budgetKrw: z.number().int().positive().nullish(),
    tripPurpose: z.string().nullish(),
    preferences: z.array(z.string()).optional(),
    selectedHotelId: z.string().nullish(),
    approvalStatus: approvalStatusSchema.optional(),
    lastUserAction: z.string().nullish(),
    pendingAction: z.string().nullish(),
    requirementCards: z.array(requirementCardSchema).optional(),
    draftPlan: draftPlanSchema.nullish(),
    hotelOptions: z.array(hotelOptionSchema).optional(),
    policyChecks: z.array(policyCheckSchema).optional(),
    finalPlan: finalPlanSchema.nullish(),
  })
  .passthrough();

export type RequirementCardState = z.infer<typeof requirementCardSchema>;
export type DraftPlanState = z.infer<typeof draftPlanSchema>;
export type HotelOptionState = z.infer<typeof hotelOptionSchema>;
export type PolicyCheckState = z.infer<typeof policyCheckSchema>;
export type FinalPlanState = z.infer<typeof finalPlanSchema>;
export type BusinessTripWorkspaceState = z.infer<
  typeof businessTripWorkspaceStateSchema
>;
export type BusinessTripWorkspacePatch = z.infer<
  typeof businessTripWorkspacePatchSchema
>;

export function emptyBusinessTripWorkspaceState(
  stage: BusinessTripStage,
): BusinessTripWorkspaceState {
  return {
    stage,
    workflowStage: stage,
    approvalStatus: "not_requested",
    preferences: [],
    requirementCards: [],
    hotelOptions: [],
    policyChecks: [],
  };
}

export function compactBusinessTripPatch(
  patch: BusinessTripWorkspacePatch,
): BusinessTripWorkspacePatch {
  return Object.fromEntries(
    Object.entries(patch).filter(([, value]) => value !== undefined),
  ) as BusinessTripWorkspacePatch;
}

export function coerceBusinessTripWorkspaceState(
  value: unknown,
  fallbackStage: BusinessTripStage,
): BusinessTripWorkspaceState {
  const parsed = businessTripWorkspaceStateSchema.safeParse(value);

  if (!parsed.success) {
    return emptyBusinessTripWorkspaceState(fallbackStage);
  }

  const parsedStage = parsed.data.stage;
  const stage =
    isBusinessTripStage(parsedStage)
      ? parsedStage
      : parsed.data.workflowStage ?? fallbackStage;

  return {
    ...parsed.data,
    stage,
    workflowStage: parsed.data.workflowStage ?? stage,
    approvalStatus: parsed.data.approvalStatus ?? "not_requested",
    requirementCards: parsed.data.requirementCards ?? [],
    preferences: parsed.data.preferences ?? [],
    hotelOptions: parsed.data.hotelOptions ?? [],
    policyChecks: parsed.data.policyChecks ?? [],
  };
}

export function mergeBusinessTripWorkspaceState(
  currentState: BusinessTripWorkspaceState,
  patch: BusinessTripWorkspacePatch,
): BusinessTripWorkspaceState {
  const compactPatch = compactBusinessTripPatch(patch);
  const nextStage =
    compactPatch.stage ??
    compactPatch.workflowStage ??
    (isBusinessTripStage(currentState.stage) ? currentState.stage : "requirements");

  return {
    ...currentState,
    ...compactPatch,
    stage: nextStage,
    workflowStage: compactPatch.workflowStage ?? nextStage,
    approvalStatus:
      compactPatch.approvalStatus ?? currentState.approvalStatus ?? "not_requested",
    requirementCards:
      compactPatch.requirementCards ?? currentState.requirementCards ?? [],
    preferences: compactPatch.preferences ?? currentState.preferences ?? [],
    hotelOptions: compactPatch.hotelOptions ?? currentState.hotelOptions ?? [],
    policyChecks: compactPatch.policyChecks ?? currentState.policyChecks ?? [],
  };
}

export function formatKrw(value: number | null | undefined) {
  if (value == null) {
    return "";
  }

  return `${new Intl.NumberFormat("ko-KR").format(value)}원`;
}
